src/Symfony/Component/Mapper/Tests/MapTest.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace Symfony\Component\Mapper\Tests;
+
+use PHPUnit\Framework\TestCase;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Mapper\AutoMapper;
+use Symfony\Component\Mapper\Exception\RuntimeException;
+use Symfony\Component\Mapper\Tests\Fixtures\A;
+use Symfony\Component\Mapper\Tests\Fixtures\B;
+use Symfony\Component\Mapper\Tests\Fixtures\C;
+use Symfony\Component\Mapper\Tests\Fixtures\D;
+use Symfony\Component\Mapper\Tests\Fixtures\DeeperRecursion\Recursive;
+use Symfony\Component\Mapper\Tests\Fixtures\DeeperRecursion\RecursiveDto;
+use Symfony\Component\Mapper\Tests\Fixtures\DeeperRecursion\Relation;
+use Symfony\Component\Mapper\Tests\Fixtures\DeeperRecursion\RelationDto;
+use Symfony\Component\Mapper\Tests\Fixtures\MultipleTargets\A as MultipleTargetsA;
+use Symfony\Component\Mapper\Tests\Fixtures\MultipleTargets\C as MultipleTargetsC;
+use Symfony\Component\Mapper\Tests\Fixtures\Recursion\Dto;
+use Symfony\Component\Mapper\Tests\Fixtures\Recursion\AB;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+
+final class MapTest extends TestCase
+{
+    /**
+     * @dataProvider mapProvider
+     */
+    public function testMap($expect, $args, array $deps = []): void
+    {
+        $mapper = new AutoMapper(...$deps);
+        $this->assertEquals($expect, $mapper->map(...$args));
+    }
+
+    /**
+     * @return array{expect: object, args: array, deps: array}
+     */
+    public function mapProvider()
+    {
+        $d = new D(baz: 'foo', bat: 'bar');
+        $c = new C(foo: 'foo', bar: 'bar');
+        $a = new A();
+        $a->foo = 'test';
+        $a->transform = 'test';
+        $a->baz = 'me';
+        $a->notinb = 'test';
+        $a->relation = $c;
+        $a->relationNotMapped = $d;
+
+        $b = new B('test');
+        $b->transform = 'TEST';
+        $b->baz = 'me';
+        $b->nomap = true;
+        $b->concat = 'testme';
+        $b->relation = $d;
+        $b->relationNotMapped = $d;
+        yield [$b, [$a]];
+
+        $c = clone $b;
+        $c->id = 1;
+        yield [$c, [$a, $c]];
+
+        $d = clone $b;
+        // with propertyAccessor we call the getter
+        $d->concat = 'shouldtestme';
+
+        yield [$d, [$a], [PropertyAccess::createPropertyAccessor()]];
+
+        $e = clone $b;
+        $e->transform = 'Test';
+        $serviceLocator = $this->createMock(ContainerInterface::class);
+        $serviceLocator->method('has')->willReturnCallback(function ($v): bool {
+            return $v === 'strtoupper';
+        });
+        $serviceLocator->method('get')->with('strtoupper')->willReturn(fn ($v) => ucfirst($v));
+
+        yield [$e, [$a], [null, $serviceLocator]];
+
+        yield [new MultipleTargetsC(), [new MultipleTargetsA()]];
+    }
+
+    public function testHasNothingToMapTo(): void
+    {
+        $this->expectException(RuntimeException::class);
+        (new AutoMapper)->map(new class {});
+    }
+
+    public function testRecursion(): void {
```

Comment:
we don't use `void` in tests
```suggestion
    public function testRecursion()
    {
```